test/unit/ServerClientErrors.test.ts

```diff
@@ -30,40 +30,89 @@ describe("ServerClient - Errors", () => {
-                expect(error.message).to.equal("Basic error");
+                return client.getServer().then((result) => {
+                    return result;
+                }, (error) => {
+                    expect(error).to.be.instanceOf(postmark.Errors.PostmarkError);
+                });
+            });
+
+            it("message", () => {
+                sandbox.stub(BaseClient.prototype, <any> "httpRequest").rejects({ message: "Basic error", response: {data: "Basic error" }});
+
+                const serverToken: string = testingKeys.get("SERVER_TOKEN");
+                let client: postmark.ServerClient = new postmark.ServerClient(serverToken);
+
+                return client.getServer().then((result) => {
+                    return result;
+                }, (error) => {
+                    expect(error.message).to.equal("Basic error");
+                });
+            });
+
+            it("name", () => {
+                sandbox.stub(BaseClient.prototype, <any> "httpRequest").rejects({ response: { data: 'response', status: 401} });
+
+                return client.getBounces().then((result) => {
+                    return result;
+                }, (error) => {
+                    expect(error.name).to.equal(invalidTokenError);
+                });
             });
         });
 
-        it("throw api key error - promise", () => {
-            let error: any = new Error("Basic error");
-            error.statusCode = 401;
-            sandbox.stub(BaseClient.prototype, <any> "httpRequest").throws(error);
+        describe("callback error", () => {
+            it("name", (done) => {
+                sandbox.stub(BaseClient.prototype, <any> "httpRequest").rejects({ response: {data: "Basic error", status: 404}});
 
-            const serverToken: string = testingKeys.get("SERVER_TOKEN");
-            let client: postmark.ServerClient = new postmark.ServerClient(serverToken);
+                const serverToken: string = testingKeys.get("SERVER_TOKEN");
+                let client: postmark.ServerClient = new postmark.ServerClient(serverToken);
 
-            return client.getServer().then((result) => {
-                return result;
-            }, (error) => {
-                expect(error).to.be.instanceOf(postmark.Errors.InvalidAPIKeyError);
+                client.getServer((error: any, data) => {
+                    expect(data).to.equal(null);
+                    expect(error.name).to.equal('PostmarkError');
+                    done();
+                });
             });
         });
 
+        it("empty token", () => {
+            expect(() => new postmark.ServerClient(""))
+                .to.throw("A valid API token must be provided.");
+        });
+
         describe("http status code errors", () => {
+            let error: any = {
+                response: {
+                    data: "Basic error",
+                    status: 505
+                }
+            };
```

Comment:
You need to be careful here - a variable defined at this level leaks between tests and sometimes can lead to subtle bugs. It would be nice to have a similar pattern here as we have in Rspec but that _let_ works in a totally different way (it's guaranteed to be isolated between tests).

A safer option would be to have a helper:

```ts
const buildError = (status: number) => ({
    response: {
        data: "Basic error",
        status,
    },
});
```

and use it like:

```ts
 sandbox.stub(BaseClient.prototype, <any> "httpRequest").rejects(buildError(401));
```

Reply:
absolutely agreed, that was a product of pure laziness, fixed, thanks!